### PR TITLE
Add multiple phase realizations to waves

### PR DESCRIPTION
**Pull request by jtgrasb**
## Description
This PR adds a `realization` dimension to the waves DataArray (Issue #262). The number of realizations is set equal to 1 by default, but is set to 5 for tutorial 3 and 4 which have irregular waves. The `solve()` function now loops over multiple phase realizations and returns a list of the results from all optimizations.

## Type of PR
- [ ] Bug fix
- [X] New feature
- [ ] Documentation
- [ ] Other: (specify)

## Checklist for PR
- [X] Authors read the [contribution guidelines](https://github.com/sandialabs/WecOptTool/blob/main/.github/CONTRIBUTING.md)
- [X] The pull request is from an issue branch (not main) on *your* fork, to the [main branch in WecOptTool](https://github.com/sandialabs/WecOptTool).
- [X] The authors have given the admins edit access
- [X] All changes adhere to the style guide including PEP8, Docstrings, and Type Hints.
- [ ] Modified the documentation if applicable
- [ ] Modified or added a new test
- [ ] All tests pass
- [ ] [Reference or close any relevant issues](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)

## Additional details
I am still working on adding a section to one of the tutorials to demonstrate multiple phase realizations vs. a longer time window.

### CODE REVIEW

**Comment by jtgrasb** [comment]
Not sure why the tests aren't being run by Github Actions. I resolved a merge conflict and now the tests aren't running.

File: wecopttool/core.py [hunk 1]
```diff
@@ -686,15 +686,15 @@ def solve(self,
         To get the post-processed results for the
         :py:class:`wecopttool.pto.PTO`, you may call
 
-        >>> res_wec_fd, res_wec_td = wec.post_process(wec,res_opt)
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:689]: update the  example text above. Something like 

> To get the post-processed results for the WEC and PTO for a specific realization, you may call

File: wecopttool/core.py [hunk 2]
```diff
@@ -686,15 +686,15 @@ def solve(self,
...
-        >>> res_pto_fd, res_pto_td = pto.post_process(wec,res_opt)
+        >>> res_wec_fd, res_wec_td = wec.post_process(wec,res_opt[0])
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:689]:
```suggestion
        >>> realization = 0 # realization index
        >>> res_wec_fd, res_wec_td = wec.post_process(wec,res_opt[realization])
```

File: wecopttool/core.py [hunk 3]
```diff
@@ -686,15 +686,15 @@ def solve(self,
...
+        >>> res_pto_fd, res_pto_td = pto.post_process(wec,res_opt[0])
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:690]:
```suggestion
        >>> res_pto_fd, res_pto_td = pto.post_process(wec,res_opt[realization])
```

File: wecopttool/core.py [hunk 4]
```diff
@@ -791,38 +784,52 @@ def callback_scipy(x):
...
-        # minimize
-        optim_res = minimize(**problem)
+                return callback(self, x_wec, x_opt, wave)
+
+        for realization, wave in waves.groupby('realization'):
+
+            _log.info("Solving pseudo-spectral control problem "
+                      + f"for realization number {realization}.") 
+
+            # decision variable initial guess
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:794]: I don't think we need a different initial guess for each realization... I think we can keep this outside the loop

Comment (cmichelenstrofer) [reply wecopttool/core.py:794]: conversely I would move everything that uses `wave` into the loop. This includes the constraints, objective function and callback.

File: tests/test_waves.py [hunk 5]
```diff
@@ -289,7 +300,7 @@ def test_coordinates(self, elevation):
         """Test that the elevation dataArray has the correct
         coordinates.
         """
-        coordinates = ['wave_direction', 'omega', 'freq']
+        coordinates = ['wave_direction', 'omega', 'freq', 'realization']
```
Comment (cmichelenstrofer) [inline tests/test_waves.py:303]: should similarly add a test for number of realizations like you did for the long crested waves.

Comment (cmichelenstrofer) [reply tests/test_waves.py:303]: Also below, add a test to `class TestRandomPhase` that uses `nrealizations=2`

File: tests/test_integration.py
Comment (cmichelenstrofer) [inline tests/test_integration.py:1]: The `irregular_wave` is never used... The function `test_unstructured_controller_irregular_wave` uses `regular_wave`. @ryancoe I think this is a bug. @jtgrasb this mistake was not introduced by your PR but I think this PR might be a good place to fix it and add `nrealizations=2` to `irregular_wave`.

File: wecopttool/waves.py [hunk 6]
```diff
@@ -185,6 +191,7 @@ def regular_wave(
 def long_crested_wave(
     efth: DataArray,
     direction: Optional[float] = 0.0,
+    nrealizations: Optional[float] = 1,
```
[the remark this answers is not in the document]
Comment (jtgrasb) [reply wecopttool/waves.py]: Made it a required input

**Pull request merged**